Documentation/design-docs/native-hosting.md

```diff
@@ -301,12 +298,69 @@ Initially there might be a limitation of calling this function only once on a gi
 ```
 Closes a host context.
 * `host_context_handle` - handle to the initialized host context to close.
 
 
+### Multiple host contexts interactions
+
+It is important to correctly synchronize some of these operations to achieve the desired API behavior as well as thread safety requirements. The following behaviors will be used to achieve this.
+
+#### Terminology
+* `first host context` is the one which is used to load and initialize the CoreCLR runtime in the process. At any given time there can only be one `first host context`.
+* `secondary host context` is any other initialized host context when `first host context` already exists in the process.
+
+#### Synchronization
+* If there's no `first host context` in the process the first call to `hostfxr_initialize_...` will create a new `first host context`. There can only be one `first host context` in existence at any point in time.
+* Calling `hostfxr_initialize...` when `first host context` already exists will always return a `secondary host context`.
+* The `first host context` blocks creation of any other host context until it is used to load and initialize the CoreCLR runtime. This means that `hostfxr_initialize...` and subsequently one of the "run" methods must be called on the `first host context` to unblock creation of `secondary host contexts`.
+* Calling `hostfxr_initialize...` will block until the `first host context` is initialized, a "run" method is called on it and the CoreCLR is loaded and initialized. The `hostfxr_initialize...` will block potentially indefinitely. The method will block very early on, all of the operations done by the initialize will only happen once it's unblocked.
+* `first host context` can fail to initialize the runtime (or anywhere up to that point). If this happens, it's marked as failed and is not considered a `first host context` anymore. This unblocks the potentially waiting `hostfxr_initialize...` calls. In this case the first `hostfxr_initialize...` after the failure will create a new `first host context`.
+* `first host context` can be closed using `hostfxr_close` before it is used to initialize the CoreCLR runtime. This is similar to the failure above, the host context is marked as "closed/failed" and is not considered `first host context` anymore. This unblocks any waiting `hostfxr_initialize...` calls.
+* Once the `first host context` successfully initialized the CoreCLR runtime it is permanently marked as "successful" and will remain the `first host context` for the lifetime of the process. Such host context can (and should) be closed via the `hostfxr_close` when not used anymore, but internally it will remain active.
+
+#### Invalid usage
+* It is invalid to initialize a host context via `hostfxr_initialize...` and then never call `hostfxr_close` on it. Initialized but not closed host context is abandoned. Abandoned `first host context` will cause infinitely blocking of any future `hostfxr_initialize...` calls.
```

Comment:
~~Initialized but not closed host context is abandoned.~~ -> An initialized but not closed `host context` is considered abandoned.

Reply:
infinitely -> infinite